src/backend/distributed/commands/database.c

```diff
@@ -242,3 +284,187 @@ PreprocessAlterDatabaseSetStmt(Node *node, const char *queryString,
 
 	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
 }
+
+
+/*
+ * PostprocessCreatedbStmt is executed after the statement is applied to the local
+ * postgres instance. In this stage we can prepare the commands that need to be run on
+ * all workers to create the database.
+ *
+ */
+List *
+PostprocessCreateDatabaseStmt(Node *node, const char *queryString)
+{
+	if (EnableCreateDatabasePropagation)
+	{
+		EnsureCoordinator();
+	}
+
+	if (!EnableCreateDatabasePropagation || !ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	CreatedbStmt *stmt = castNode(CreatedbStmt, node);
+	char *databaseName = stmt->dbname;
+	bool missingOk = false;
+	Oid databaseOid = get_database_oid(databaseName, missingOk);
+
+	/*
+	 * TODO: try to reuse regular DDL infrastructure
+	 *
+	 * We do not do this right now because of the AssignDatabaseToShard at the end.
+	 */
+	List *workerNodes = TargetWorkerSetNodeList(NON_COORDINATOR_METADATA_NODES,
+												RowShareLock);
+	if (list_length(workerNodes) > 0)
+	{
+		char *createDatabaseCommand = DeparseTreeNode(node);
+
+		StringInfo internalCreateCommand = makeStringInfo();
+		appendStringInfo(internalCreateCommand,
+						 "SELECT pg_catalog.citus_internal_database_command(%s)",
+						 quote_literal_cstr(createDatabaseCommand));
+
+		/*
+		 * For the moment, we run CREATE DATABASE in 2PC, though that prevents
+		 * us from immediately doing a pg_dump | pg_restore when dealing with
+		 * a remote template database.
+		 */
+		bool outsideTransaction = false;
+
+		List *taskList = CreateDDLTaskList(internalCreateCommand->data, workerNodes,
+										   outsideTransaction);
+
+		bool localExecutionSupported = false;
+		ExecuteUtilityTaskList(taskList, localExecutionSupported);
+	}
+
+	/* synchronize pg_dist_object records */
+	ObjectAddress dbAddress = { 0 };
+	ObjectAddressSet(dbAddress, DatabaseRelationId, databaseOid);
+	MarkObjectDistributed(&dbAddress);
+
+
+	return NIL;
+}
+
+
+/*
+ * citus_internal_database_command is an internal UDF to
+ * create/drop a database in an idempotent maner without
+ * transaction block restrictions.
+ */
+Datum
+citus_internal_database_command(PG_FUNCTION_ARGS)
```

Comment:
@marcoslot I don't understand why we use a GUC to get rid of transaction restriction. What a GUC do to remove the transaction?

Reply:
Yes :) editing the comment

Reply:
> @marcoslot I don't understand why we use a UDF to get rid of transaction restriction. What a UDF do to remove the transaction?

Reply:
@gurkanindibay, I think we can avoid from adding this function by setting `cannotBeExecutedInTransction` field to false for the `Task`s generated for CREATE/DROP database commands, as in https://github.com/citusdata/citus/commit/5fae7b9ad614e06f2205d8f08b9914deed15dea0

We should probably avoid from those ugly loops either:
i) by adding a param to NodeDDLTaskList that allows specifying whether cannotBeExecutedInTransction needs to be set to false, or,
ii) by adding a variation of NodeDDLTaskList for CREATE/DROP database commands that implicitly sets cannotBeExecutedInTransction to false

Reply:
If we decide using `cannotBeExecutedInTransaction` flag to do that, then we need to fix following:


catch block in utility hook assumes the command is an index command if `cannotBeExecutedInTransaction` is set to true for a Task, need to fix that:

https://github.com/citusdata/citus/blob/db13afaa7b77777003b7a724a01fdb0f5df481d1/src/backend/distributed/commands/utility_hook.c#L1252-L1282

```sql
# connected to worker at 9701
[local] onurctirtir@onurctirtir:9701-6920=# create database onur;
NOTICE:  Citus partially supports CREATE DATABASE for distributed databases
DETAIL:  Citus does not propagate CREATE DATABASE command to workers
HINT:  You can manually create a database and its extensions on workers.
CREATE DATABASE
Time: 163.724 ms

# connected to coordinator at 9700
[local] onurctirtir@onurctirtir:9700-6954=# set citus.enable_create_database_propagation to on;
SET
Time: 1.565 ms
[local] onurctirtir@onurctirtir:9700-6954=# create database onur;
WARNING:  CONCURRENTLY-enabled index commands can fail partially, leaving behind an INVALID index.
 Use DROP INDEX CONCURRENTLY IF EXISTS to remove the invalid index.
ERROR:  database "onur" already exists
CONTEXT:  while executing command on localhost:9701
```

Reply:
A candidate fix for that could be to change the warning:

```c
			else
			{
				ereport(WARNING,
						(errmsg("Commands that cannot be executed in a transaction "
                                "can fail partially, leaving behind an invalid "
                                "state. If this was a CREATE command, try dropping "
                                "the object with IF EXISTS syntax if possible and "
                                "retry the original command.")));
				PG_RE_THROW();
			}
```